### DISCUSSION TITLE
Solution for offline workflow? (rendering and viewing without Internet access)

### DISCUSSION THREAD

**Discussion by sijow** [quarto-dev/quarto-cli#9559] [Q&A]
### Description

Is there a way to fully render a Quarto file with math while being offline?

For example rendering the following with `quarto render math.qmd`

```
---
embed-resources: true
self-contained-math: true
---

$$
x^2 = 4
$$
```

shows the following messages:

```
[WARNING] Could not fetch resource https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6: HttpExceptionRequest Request {
    host                 = "cdnjs.cloudflare.com"
    port                 = 443
    secure               = True
    requestHeaders       = []
    path                 = "/polyfill/v3/polyfill.min.js"
    queryString          = "?features=es6"
    method               = "GET"
    proxy                = Nothing
    rawBody              = False
    redirectCount        = 10
    responseTimeout      = ResponseTimeoutDefault
    requestVersion       = HTTP/1.1
    proxySecureMode      = ProxySecureWithConnect
  }
   (ConnectionFailure Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [AI_ADDRCONFIG], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = 0.0.0.0:0, addrCanonName = Nothing}, host name: Just "cdnjs.cloudflare.com", service name: Just "443"): does not exist (Try again))
[WARNING] Could not fetch resource https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js: HttpExceptionRequest Request {
    host                 = "cdn.jsdelivr.net"
    port                 = 443
    secure               = True
    requestHeaders       = []
    path                 = "/npm/mathjax@3/es5/tex-chtml-full.js"
    queryString          = ""
    method               = "GET"
    proxy                = Nothing
    rawBody              = False
    redirectCount        = 10
    responseTimeout      = ResponseTimeoutDefault
    requestVersion       = HTTP/1.1
    proxySecureMode      = ProxySecureWithConnect
  }
   (ConnectionFailure Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [AI_ADDRCONFIG], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = 0.0.0.0:0, addrCanonName = Nothing}, host name: Just "cdn.jsdelivr.net", service name: Just "443"): does not exist (Try again))
Output created: math.html
```

and opening the HTML page I see `\[ x^2 = 4 \]` instead of rendered math.

I'm thinking maybe a way to have quarto use a local cache for the CDN resources, that could be populated once while online and use thereafter?

**Comment by mcanouil** [comment]
WHich format are you using? And which version of Quarto?

There is a known issue for `revealjs`:

- https://github.com/quarto-dev/quarto-cli/issues/9404

**Reply by mcanouil** [reply]
You are trying to embed something in your output that you don't have.

You need to [download MathJax](https://github.com/mathjax/MathJax/archive/master.zip) (or any library you want to use for math prior to render offline), then:

```yaml
html-math-method:
  method: mathjax
  url: path/to/mathjax.js
```

This is given to Pandoc (https://pandoc.org/MANUAL.html#math-rendering-in-html).
I am not enterily sure `url` allows path.

Edit: it does work but there is still the polyfill warning which comes from the Pandoc template (hardcoded in it) when using MathJax

- https://github.com/quarto-dev/quarto-cli/blob/84ee8099cb1631d548f68bb1876c3c5285f8afa0/src/resources/formats/html/pandoc/html.template#L30-L32
- https://github.com/jgm/pandoc/blob/main/data/templates/default.html5#L30-L32

`mathjax.js` is from <https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js>.

````qmd
---
format: html
embed-resources: true
self-contained-math: true
html-math-method:
  method: mathjax
  url: ./mathjax.js
---

$$
x^2 = 4
$$
````

**Reply by sijow** [reply]
Thanks for investigating. The proposed solution doesn't seem to work for me (quarto 1.4.554). Apart from the polyfill warning, there's a message `[WARNING] Could not fetch resource ./mathjax.js` during rendering, and the HTML still displays as `\[ x^2 = 4 \]` instead of MathJax.  Also if it did work it would probably defeat the `embed-resources: true`, `self-contained-math: true`, since the HTML output refers to `./mathjax.js` instead of embedding it.

It's indeed related to #9404. I think #9404 is about showing an HTML file while offline (the steps to reproduce mention running `quarto render` normally, then disconnecting and looking at the HTML output while offline). I'd like to have also the `quarto render` part work offline.

**Reply by mcanouil** [reply]
Did you downloaded the file first?
I literally downloaded the CDN JS file as `./mathjax.js` and used that.
I used Quarto development version but don't believe it matters here.

The JS script is properly embedded in the HTML file as it should.
For polyfill, the only options are to [modify the template](https://quarto.org/docs/journals/templates.html) or use one of the other Math libraries.

```txt
quarto preview /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd --no-browser --no-watch-inputs
Check file:///Users/mcanouil/Projects/quarto/quarto-cli/src/quarto.ts
pandoc 
  to: html
  output-file: index.html
  standalone: true
  embed-resources: true
  section-divs: true
  html-math-method:
    method: mathjax
    url: ./mathjax.js
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  
[WARNING] Could not fetch resource https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6: HttpExceptionRequest Request {
    host                 = "cdnjs.cloudflare.com"
    port                 = 443
    secure               = True
    requestHeaders       = []
    path                 = "/polyfill/v3/polyfill.min.js"
    queryString          = "?features=es6"
    method               = "GET"
    proxy                = Nothing
    rawBody              = False
    redirectCount        = 10
    responseTimeout      = ResponseTimeoutDefault
    requestVersion       = HTTP/1.1
    proxySecureMode      = ProxySecureWithConnect
  }
   (ConnectionFailure Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [AI_ADDRCONFIG], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = 0.0.0.0:0, addrCanonName = Nothing}, host name: Just "cdnjs.cloudflare.com", service name: Just "443"): does not exist (nodename nor servname provided, or not known))
Output created: index.html

Watching files for changes
Browse at http://localhost:6143/
  /output/chtml/fonts/woff-v2/MathJax_Zero.woff (404: Not Found)
  /output/chtml/fonts/woff-v2/MathJax_Math-Italic.woff (404: Not Found)
  /output/chtml/fonts/woff-v2/MathJax_Main-Regular.woff (404: Not Found)
```

The fonts not found only appears to be an issue when previewing.

**Reply by sijow** [reply]
Sorry, please ignore the `Could not fetch resource` error mentioned above: I had forgotten to rename `tex-chtml-full.js` to `mathjax.js`. 

I can get the MathJax rendering with `url: ./mathjax.js` but the math font is wrong. For example the following document

```
---
format: html
html-math-method:
  method: mathjax
  url: ./mathjax.js
---

$$
\mathcal{X} \in \mathbb{R}
$$
```

renders as
![image](https://github.com/quarto-dev/quarto-cli/assets/2412819/1dff7b1c-5c54-475f-b285-8aaf1fc3df8e)

while after commenting out the `url:` line I get the correct rendering:
![image](https://github.com/quarto-dev/quarto-cli/assets/2412819/a6513cec-099a-4ea4-974c-c98c79948b61)


Note that this font problem can also occur when I'm not using the `url:` option, for example the following document also gives the wrong font:

```
---
format: html
embed-resources: true
self-contained-math: true
---

$$
\mathcal{X} \in \mathbb{R}
$$
```

**Reply by mcanouil** [reply]
Maybe related to polyfill or to the missing fonts warnings.

**Reply by cscheid** [reply]
Another option here might be to use MathML. Its support is not perfect in many browsers, but it's been improving, and it's acceptably good in Firefox.

**Reply by sijow** [reply]
Ah yes MathML looks like a promising solution, especially now that Chrome finally supports it. I'll try and see how well it works.

**Reply by O957** [reply]
Any updates on this? Specifically, on fixing of fonts path errors. E.g.

> output/chtml/fonts/woff-v2/MathJax_Zero.woff (404: Not Found)

**Comment by likanzhan** [comment]
Any update on this?

**Reply by cderv** [reply]
Did you try recently? Polyfill is no longer used. So the OP is now outdated. 

With mathjax, there still maybe font problem as they are probably not embded - this would still require download mathjax, and setting the url to local; 

Using `html-math-method: katex` may not have this font issue; 

If you have a new issue with this, please share details. Thank you

**Reply by likanzhan** [reply]
I guess that my problem is related to bibliography ?

````
Could not fetch http://www.zotero.org/styles/multidisciplinary-digital-publishing-institute
HttpExceptionRequest Request {
  host                 = "www.zotero.org"
  port                 = 80
  secure               = False
  requestHeaders       = []
  path                 = "/styles/multidisciplinary-digital-publishing-institute"
  queryString          = ""
  method               = "GET"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
  proxySecureMode      = ProxySecureWithConnect
}
 (ConnectionFailure Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = 0.0.0.0:0, addrCanonName = Nothing}, host name: "www.zotero.org", service name: "80"): does not exist (nodename nor servname provided, or not known))
````

Thanks

**Reply by mcanouil** [reply]
Please open a new discussion and share a small and simple reproducible example following the guidelines provided when opening a new Q&A discussion.
Thanks.

**Reply by cderv** [reply]
> I guess that my problem is related to bibliography 

yes probably. Please open a new Question as asked. 

General comment though: Doing a `embed-resources: true` document does not mean you can't have internet access at the place where you render. This is meant to produce an output with only local resources. But if your resources are online at the time of rendering, Quarto (and pandoc) will require internet access to download them first before being able to embed them. 

The error you see is a Pandoc error, and it probably means 
- (a) you have your CSL file online 
- (b) you can't access internet for some reason and this CSL can't be retrieved as expected. 

Hope it helps. If not enough, please open a new topic. 

Thanks !

**Reply by likanzhan** [reply]
I created a discussion #12748
